### ISSUE TITLE
[Modules] Better support inversion of control between modules

### ISSUE THREAD

**Issue by ulrichb** [nestjs/nest#3437]
## Is your feature request related to a problem? Please describe.

Take the following simple module dependency example. When some services within `DomainModule` need to inject _implementations_ from `DataAccessModule`, then Nest forces us to do a `imports: [DataAccessModule]` in `DomainModule` (even when it's already imported by `AppModule`).

```plain

                  AppModule
                      +
                      |
      +--------------------------------+
      |                                |
      v                                v
 DomainModule +--------------> DataAccessModule
```

This is bad because there should not be any dependency from the domain layer to the "data access" (in DDD lingo "infrastructure") layer. (Instead there should only be dependencies to abstractions (interfaces) within the domain layer and `DataAccessModule` implements/registers them.)

So I want this:

```plain
                  AppModule
                      +
                      |
      +--------------------------------+
      |                                |
      v                                v
 DomainModule                  DataAccessModule
```

 ... **but** still want to inject data access _abstractions_ within `DomainModule`, which are registered within `DataAccessModule`. This module registration _must_ then happen within `AppModule`.

## Describe the solution you'd like

_Proposal:_ Allow to tag modules via a string. And allow to specify these strings as `imports` dependencies.

In the above example: Tag the `DataAccessModule` with `"DataAccessImplementations"` and allow to import this in `DomainModule` via `imports: ["DataAccessImplementations"]`.

In other words: Allow "soft dependencies" between modules.

**Comment by kamilmysliwiec** [comment]
> but still want to inject data access abstractions within DomainModule, which are registered within DataAccessModule. This module registration must then happen within AppModule.

Strict relations between modules is the key idea of the modules system in NestJS. We don't plan to change this.

**Comment by ulrichb** [comment]
While "strict relations" (for the sake of argument, let's name them strict _dependencies_) between modules is a good idea for "internal" module graphs (e.g. internal within a layer), it impedes applying the [Dependency **inversion** principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle). This is what I tried to explain in my sample above, where the "domain" layer should _not_ have a (strict) dependency to the "data access" layer.

And to avoid misunderstandings, I don't want to replace the "strict relations", I want to _add_ a "soft relation" alternative.

So maybe let's wait if more people request this possibility and revisit this in the future.

PS: Just as an example, the .NET [Autofac](https://autofac.org/) container does not have dependencies between modules (which is pretty the same concept as here) at all, neither strict nor soft (at least in the last version I've used). While this is also a draw back (I've often missed this feature) it naturally allows applying the dependency inversion principle.

**Comment by ulrichb** [comment]
Oh, I just saw that there is one more possible misunderstanding.

These "soft dependencies" don't mean that they aren't validateable.  For example if the `DomainModule` has a soft dependency to `"DataAccessImplementations"` and nobody registers the `DataAccessModule`, then this case can (and should) be detected during the start up and emit an error saying:

_"DomainModule requires some module providing the 'DataAccessImplementations' (i.e. tagged with 'DataAccessImplementations'), but nothing has been found"_.

... so the only difference of soft dependencies is that they express a dependency to a "module contract" (specified by a string token) instead of _one explicit module_ (strict dependency).
